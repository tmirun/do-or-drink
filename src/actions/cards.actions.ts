export const ADD_CARD_ACTION = 'ADD_CARD_ACTION';
export const EDIT_CARD_CATION = 'EDIT_CARD_CATION';

interface AddCard {
  type: typeof ADD_CARD_ACTION,
  payload: string;
}

interface EditCard {
  type: typeof EDIT_CARD_CATION,
  payload: {
    index: number;
    text: string;
  }
}

export type CardAction = AddCard | EditCard;

export function addCard(text: string): AddCard {
  return {
    type: ADD_CARD_ACTION,
    payload: text
  }
}

export function editCard(index: number,text: string): EditCard {
  return {
    type: EDIT_CARD_CATION,
    payload: { index, text}
  }
}