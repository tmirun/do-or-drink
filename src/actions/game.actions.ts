import { CardStatus } from './../components/Card/Card';
export const SELECT_CARD = 'SELECT_CARD';
export const INCLEMENT_COUNT = 'INCLEMENT_COUNT';
export const SET_CARD_STATUS = 'SET_CARD_STATUS';

interface SelectRandomCard {
  type: typeof SELECT_CARD
  payload: {
    card: string,
    index: number
  }
}

interface InclementCount {
  type: typeof INCLEMENT_COUNT,
}

interface SetCardStatus {
  type: typeof SET_CARD_STATUS,
  payload: CardStatus
}

export type GameAction = SelectRandomCard | InclementCount | SetCardStatus;

export function selectRandomCard(cards: string[]): SelectRandomCard {
  const index = Math.floor(Math.random() * cards.length);
  const card = cards[index];
  return {
    type: SELECT_CARD,
    payload: { card, index }
  }
}

export function inclementCount(index: number,text: string): InclementCount {
  return {
    type: INCLEMENT_COUNT
  }
}