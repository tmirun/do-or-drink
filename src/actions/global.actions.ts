import { AppState } from './../reducers/index';
export const REFRESH_ALL_STATE = 'REFRESH_ALL_STATE';

interface RefreshAllState {
  type: typeof REFRESH_ALL_STATE,
  payload: AppState;
}

export type GlobalAction = RefreshAllState;

export function refreshAllState(newState: AppState): GlobalAction {
  return {
    type: REFRESH_ALL_STATE,
    payload: newState
  }
}