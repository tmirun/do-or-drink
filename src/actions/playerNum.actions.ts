export const SET_PLAYER_NUM_ACTION = 'SET_PLAYER_NUM_ACTION';

interface SetPlayerNumActionType {
  type: typeof SET_PLAYER_NUM_ACTION,
  payload: number;
}

export type PlayerNumActionType = SetPlayerNumActionType;

export function setPlayerNum(num: number): SetPlayerNumActionType {
  return {
    type: SET_PLAYER_NUM_ACTION,
    payload: num
  }
}