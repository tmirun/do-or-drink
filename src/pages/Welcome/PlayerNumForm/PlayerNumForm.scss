@import '../../../scss/variables.scss';
@import '~materialize-css/sass/materialize.scss';

.PlayerNumForm {
  padding: $spacing-m;
  display: flex;
  flex-direction: column;

  input {
    text-align: center;
  }

  &__error {
    @extend .red-text, .text-lighten-1;
    margin-top: $spacing-s;
  }

  button {
    margin-top: $spacing-m;
  }
}