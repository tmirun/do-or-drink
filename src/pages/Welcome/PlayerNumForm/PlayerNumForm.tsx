import React from 'react'
import { Formik, FormikProps, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import { setPlayerNum } from '../../../actions/playerNum.actions';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router';
import { ROUTE_PREPARATION_PATH } from '../../../AppRoutes';

import './PlayerNumForm.scss';

interface PlayerNumFormValues {
  playerNum: number;
}

export const PlayerNumForm: React.FC = () => {
  const dispatch = useDispatch()
  const history = useHistory();

  const initValues: PlayerNumFormValues = { playerNum: 0 }

  const onSubmit = (values: PlayerNumFormValues) => {
    dispatch(setPlayerNum(values.playerNum));
    history.push(ROUTE_PREPARATION_PATH);
  }
  return(
    <Formik
      initialValues={initValues}
      onSubmit={onSubmit}
      validationSchema = { Yup.object().shape({
        playerNum: Yup.number().required().moreThan(2, 'Player must be more than 2')
      })}
      render={(props: FormikProps<PlayerNumFormValues>) => (
        <form className="PlayerNumForm" onSubmit={(props.handleSubmit)}>
          <Field type="number" name="playerNum" placeholder="Player number"/>
          <ErrorMessage name="playerNum">
            {msg => <div className='PlayerNumForm__error'>{msg}</div>}
          </ErrorMessage>
          <button className="btn" type="submit"> next &rarr; </button>
        </form>
      )}
    />
  );
}