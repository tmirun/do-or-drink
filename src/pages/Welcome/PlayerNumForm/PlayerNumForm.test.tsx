import React from "react";
import { mount, ReactWrapper } from 'enzyme';
import { PlayerNumForm } from './PlayerNumForm';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import wait from "waait";
import { SET_PLAYER_NUM } from "../../../actions/playerNum.actions";

const mockStore = configureStore([])

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => {},
}));

describe('PlayerNumForm', () => {
  const initialState = {}
  let store: any;
  let wrapper: ReactWrapper;

  beforeEach(() => {
    store = mockStore(initialState);

    wrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <PlayerNumForm/>
        </BrowserRouter>
      </Provider>
    )
  })
  
  it('should render correctlly', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should show error message if user is less than 2', async () => {
    const inputEle = wrapper.find('input');

    inputEle.simulate('change', { 
      target: { name: 'playerNum', value: 0 },
    });
    inputEle.simulate("blur");
    await wait(0);
    wrapper.update();

    expect(wrapper.exists('.PlayerNumForm__error')).toEqual(true);
  }) 

  it('should hide error message if user is less than 2', async () => {
    const inputEle = wrapper.find('input');

    inputEle.simulate('change', { 
      target: { name: 'playerNum', value: 3 },
    });
    inputEle.simulate('blur');
    await wait(10);
    wrapper.update();

    expect(wrapper.exists('.PlayerNumForm__error')).toEqual(false);
  }) 

  it('should show trigger SET_PLAYER_NUM action', async () => {
    const value = 4; 
    const inputEle = wrapper.find('input');
    const formEle = wrapper.find('form');

    inputEle.simulate('change', { target: { name: 'playerNum', value }});
    inputEle.simulate("blur");
    formEle.simulate('submit');
    await wait(0);
    wrapper.update();
    
    const actions = store.getActions();
    expect(actions[0]).toEqual({ type: SET_PLAYER_NUM, payload: value })
  });
});
