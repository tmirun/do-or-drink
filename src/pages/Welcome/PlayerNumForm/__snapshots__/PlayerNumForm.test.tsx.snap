// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerNumForm should render correctlly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <PlayerNumForm>
        <Formik
          enableReinitialize={false}
          initialValues={
            Object {
              "playerNum": 0,
            }
          }
          isInitialValid={false}
          onSubmit={[Function]}
          render={[Function]}
          validateOnBlur={true}
          validateOnChange={true}
          validationSchema={
            ObjectSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_defaultDefault": [Function],
              "_deps": Array [],
              "_excludedEdges": Array [],
              "_exclusive": Object {},
              "_mutate": undefined,
              "_nodes": Array [
                "playerNum",
              ],
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "object",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "fields": Object {
                "playerNum": NumberSchema {
                  "_blacklist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "_conditions": Array [],
                  "_deps": Array [],
                  "_exclusive": Object {
                    "min": true,
                    "required": true,
                  },
                  "_mutate": undefined,
                  "_options": Object {
                    "abortEarly": true,
                    "recursive": true,
                  },
                  "_type": "number",
                  "_typeError": [Function],
                  "_whitelist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "tests": Array [
                    [Function],
                    [Function],
                  ],
                  "transforms": Array [
                    [Function],
                  ],
                },
              },
              "tests": Array [],
              "transforms": Array [
                [Function],
              ],
            }
          }
        >
          <form
            className="PlayerNumForm"
            onSubmit={[Function]}
          >
            <FormikConnect(FieldInner)
              name="playerNum"
              placeholder="Player number"
              type="number"
            >
              <FieldInner
                formik={
                  Object {
                    "dirty": false,
                    "errors": Object {},
                    "handleBlur": [Function],
                    "handleChange": [Function],
                    "handleReset": [Function],
                    "handleSubmit": [Function],
                    "initialValues": Object {
                      "playerNum": 0,
                    },
                    "isSubmitting": false,
                    "isValid": false,
                    "isValidating": false,
                    "registerField": [Function],
                    "resetForm": [Function],
                    "setError": [Function],
                    "setErrors": [Function],
                    "setFieldError": [Function],
                    "setFieldTouched": [Function],
                    "setFieldValue": [Function],
                    "setFormikState": [Function],
                    "setStatus": [Function],
                    "setSubmitting": [Function],
                    "setTouched": [Function],
                    "setValues": [Function],
                    "status": undefined,
                    "submitCount": 0,
                    "submitForm": [Function],
                    "touched": Object {},
                    "unregisterField": [Function],
                    "validate": undefined,
                    "validateField": [Function],
                    "validateForm": [Function],
                    "validateOnBlur": true,
                    "validateOnChange": true,
                    "validationSchema": ObjectSchema {
                      "_blacklist": RefSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_conditions": Array [],
                      "_defaultDefault": [Function],
                      "_deps": Array [],
                      "_excludedEdges": Array [],
                      "_exclusive": Object {},
                      "_mutate": undefined,
                      "_nodes": Array [
                        "playerNum",
                      ],
                      "_options": Object {
                        "abortEarly": true,
                        "recursive": true,
                      },
                      "_type": "object",
                      "_typeError": [Function],
                      "_whitelist": RefSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "fields": Object {
                        "playerNum": NumberSchema {
                          "_blacklist": RefSet {
                            "list": Set {},
                            "refs": Map {},
                          },
                          "_conditions": Array [],
                          "_deps": Array [],
                          "_exclusive": Object {
                            "min": true,
                            "required": true,
                          },
                          "_mutate": undefined,
                          "_options": Object {
                            "abortEarly": true,
                            "recursive": true,
                          },
                          "_type": "number",
                          "_typeError": [Function],
                          "_whitelist": RefSet {
                            "list": Set {},
                            "refs": Map {},
                          },
                          "tests": Array [
                            [Function],
                            [Function],
                          ],
                          "transforms": Array [
                            [Function],
                          ],
                        },
                      },
                      "tests": Array [],
                      "transforms": Array [
                        [Function],
                      ],
                    },
                    "values": Object {
                      "playerNum": 0,
                    },
                  }
                }
                name="playerNum"
                placeholder="Player number"
                type="number"
              >
                <input
                  name="playerNum"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Player number"
                  type="number"
                  value={0}
                />
              </FieldInner>
            </FormikConnect(FieldInner)>
            <FormikConnect(ErrorMessageImpl)
              name="playerNum"
            >
              <ErrorMessageImpl
                formik={
                  Object {
                    "dirty": false,
                    "errors": Object {},
                    "handleBlur": [Function],
                    "handleChange": [Function],
                    "handleReset": [Function],
                    "handleSubmit": [Function],
                    "initialValues": Object {
                      "playerNum": 0,
                    },
                    "isSubmitting": false,
                    "isValid": false,
                    "isValidating": false,
                    "registerField": [Function],
                    "resetForm": [Function],
                    "setError": [Function],
                    "setErrors": [Function],
                    "setFieldError": [Function],
                    "setFieldTouched": [Function],
                    "setFieldValue": [Function],
                    "setFormikState": [Function],
                    "setStatus": [Function],
                    "setSubmitting": [Function],
                    "setTouched": [Function],
                    "setValues": [Function],
                    "status": undefined,
                    "submitCount": 0,
                    "submitForm": [Function],
                    "touched": Object {},
                    "unregisterField": [Function],
                    "validate": undefined,
                    "validateField": [Function],
                    "validateForm": [Function],
                    "validateOnBlur": true,
                    "validateOnChange": true,
                    "validationSchema": ObjectSchema {
                      "_blacklist": RefSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_conditions": Array [],
                      "_defaultDefault": [Function],
                      "_deps": Array [],
                      "_excludedEdges": Array [],
                      "_exclusive": Object {},
                      "_mutate": undefined,
                      "_nodes": Array [
                        "playerNum",
                      ],
                      "_options": Object {
                        "abortEarly": true,
                        "recursive": true,
                      },
                      "_type": "object",
                      "_typeError": [Function],
                      "_whitelist": RefSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "fields": Object {
                        "playerNum": NumberSchema {
                          "_blacklist": RefSet {
                            "list": Set {},
                            "refs": Map {},
                          },
                          "_conditions": Array [],
                          "_deps": Array [],
                          "_exclusive": Object {
                            "min": true,
                            "required": true,
                          },
                          "_mutate": undefined,
                          "_options": Object {
                            "abortEarly": true,
                            "recursive": true,
                          },
                          "_type": "number",
                          "_typeError": [Function],
                          "_whitelist": RefSet {
                            "list": Set {},
                            "refs": Map {},
                          },
                          "tests": Array [
                            [Function],
                            [Function],
                          ],
                          "transforms": Array [
                            [Function],
                          ],
                        },
                      },
                      "tests": Array [],
                      "transforms": Array [
                        [Function],
                      ],
                    },
                    "values": Object {
                      "playerNum": 0,
                    },
                  }
                }
                name="playerNum"
              />
            </FormikConnect(ErrorMessageImpl)>
            <button
              className="btn"
              type="submit"
            >
               next → 
            </button>
          </form>
        </Formik>
      </PlayerNumForm>
    </Router>
  </BrowserRouter>
</Provider>
`;
